File: java/amp.tests.integration/src/test/java/amp/tests/integration/DefaultCommandBusTests.java
package amp.tests.integration;

import java.util.Map;
import java.util.concurrent.TimeUnit;

import org.junit.*;
import static org.junit.Assert.*;
import static com.jayway.awaitility.Awaitility.*;
import static org.hamcrest.Matchers.*;

import org.springframework.context.support.FileSystemXmlApplicationContext;

import amp.commanding.ICommandBus;
import amp.commanding.ICommandHandler;
import amp.messaging.MessageException;

public class DefaultCommandBusTests {
    
	protected static FileSystemXmlApplicationContext context;
	protected static ICommandBus bus;
	
	@BeforeClass
	public static void BeforeAllTests(){
		context = new FileSystemXmlApplicationContext(DefaultEventBusTests.getConfigFiles());
		bus = (ICommandBus) context.getBean("commandBus");
	}
	
	@AfterClass
	public static void AfterAllTests(){
		bus.dispose();
		context.close();
	}
	
    @Test
    public void Should_be_able_to_publish_and_subscribe() throws Exception, MessageException
    {
    	TestHandler handler = new TestHandler();
        bus.onCommandReceived(handler);
        Thread.sleep(5000);
        TestEvent sentCommand = new TestEvent();

        bus.send(sentCommand);

        waitAtMost(5, TimeUnit.SECONDS).untilCall(to(handler).getReceivedCommand(), notNullValue());
        assertEquals("Received and Sent commands were not the same.", 
        		sentCommand.Id, handler.getReceivedCommand().Id);
    }

    private class TestHandler implements ICommandHandler<TestEvent>{
    	
    	private TestEvent receivedCommand;
    	
    	public TestEvent getReceivedCommand(){
    		return receivedCommand;
    	}
    	
		public Class<TestEvent> getCommandType() {
			return TestEvent.class;
		}

		public void handle(TestEvent command, Map<String, String> headers) {
			receivedCommand = command;
		}
    }
}

File: dotnet/tests/amp.tests.integration/Commanding/DefaultCommandBusTests.cs
using System;
using System.Threading;
using amp.commanding;
using NUnit.Framework;
using Spring.Context;
using Spring.Context.Support;

namespace amp.tests.integration.Commanding
{
    [TestFixture]
    public class DefaultCommandBusTests
    {
        protected IApplicationContext _context;
        protected ICommandBus _bus;

        [TestFixtureSetUp]
        public virtual void TestFixtureSetUp()
        {
            _context = new XmlApplicationContext(Config.Bus.All, Config.Authorization.AnubisOneWaySsl, Config.Topology.GtsSSL);
            _bus = _context.GetObject("ICommandBus") as ICommandBus;
        }

        [TestFixtureTearDown]
        public virtual void TestFixtureTearDown()
        {
            _context.Dispose();
        }

        [Test]
        public void Should_be_able_to_publish_and_subscribe()
        {
            
            TestCommand receivedCommand = null;
            _bus.ReceiveCommand<TestCommand>((@command, headers) =>
            {
                receivedCommand = @command;
            });

            TestCommand sentCommand = new TestCommand();

            _bus.Send(sentCommand);

            var end = DateTime.Now.AddSeconds(5);
            while (receivedCommand == null && DateTime.Now < end)
            {
                Thread.Sleep(100);
            }

            Assert.That(receivedCommand, Is.Not.Null, "Command not received within timeout period.");
            Assert.That(receivedCommand.Id, Is.EqualTo(sentCommand.Id), "Received and Sent commands were not the same.");
        }

        public class TestCommand
        {
            public Guid Id { get; set; }

            public TestCommand()
            {
                Id = Guid.NewGuid();
            }
        }
    }
}
